README.md
# SSST_blood_sugar

Wearable device for measuring blood sugar without analysing the blood or piecing the skin.

## SDK Installation

[Video instructions](https://youtube.com/playlist?list=PLx_tBuQ_KSqEt7NK-H7Lu78lT2OijwIMl)

Windows:
1. Install nRF Command Line Tools ([Download](https://bit.ly/2YgBGC5))
2. Install nRF Connect SDK through the Toolchain Manager in nRF Connect for Desktop ([Download](https://bit.ly/39Tm3my))
3. Install Visual Studio Code ([Download](https://code.visualstudio.com/Download))
4. **Make sure to install SDK version 1.9.0!**

## Development setup

1. After cloning git repository, open `Firmware.code-workspace` with VS Code.
2. Go to nRF Connect and create *build configuration*. Under APPLICATIONS add build configuration to Firmware.
   - Board: `nrf52840dk_nrf52840`
   - Build directory name: `build`
   - Enable debug options

3. Check `Pics/nRF52840DK.jpg` for how to connect power and debugger to development kit.

For API and how to, check [Nordic's developer website](https://developer.nordicsemi.com/nRF_Connect_SDK/doc/1.9.0/nrf/index.html).

### Git commit prefixes
When committing, use prefix for easier changelog generation.

Good example: `<prefix>: <Issue title>`

Bad example: `Fixed this and this with this commit`

Prefixes:
- `enh` for enhancement (new feature)
- `bug` for bug fixes
- `hw` for HW update
- `ref` for code refactoring



Changelog.md
# Changelog

## [HW_v1](https://github.com/AZURTEST-CORP/SSST_blood_sugar/tree/HW_v1) (2022-05-23)

[Full Changelog](https://github.com/AZURTEST-CORP/SSST_blood_sugar/compare/FW_v1...HW_v1)

**Closed issues:**

- Two V1 schematics [\#18](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/18)
- Case for PCB [\#17](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/17)

**Merged pull requests:**

- HW version 1 [\#27](https://github.com/AZURTEST-CORP/SSST_blood_sugar/pull/27) ([jernejp21](https://github.com/jernejp21))

## [FW_v1](https://github.com/AZURTEST-CORP/SSST_blood_sugar/tree/FW_v1) (2022-04-29)

[Full Changelog](https://github.com/AZURTEST-CORP/SSST_blood_sugar/compare/v0...FW_v1)

**Implemented enhancements:**

- Switching between external and internal ADC [\#19](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/19)
- Implement BLE streaming [\#2](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/2)

**Fixed bugs:**

- ADC switching wile BLE not established [\#21](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/21)
- ABS value conversion not working [\#20](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/20)

**Merged pull requests:**

- Milestone V1 [\#22](https://github.com/AZURTEST-CORP/SSST_blood_sugar/pull/22) ([jernejp21](https://github.com/jernejp21))

## [v0](https://github.com/AZURTEST-CORP/SSST_blood_sugar/tree/v0) (2022-04-07)

[Full Changelog](https://github.com/AZURTEST-CORP/SSST_blood_sugar/compare/35cff00ba6ae8f5cf1055288acc48ba81a52e7a6...v0)

**Implemented enhancements:**

- External A/D converter [\#12](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/12)
- Low battery state [\#11](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/11)
- Application layer [\#7](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/7)
- System init [\#6](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/6)
- ADC switch button [\#5](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/5)
- Shutdown button [\#4](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/4)
- Program architecture [\#1](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/1)

**Fixed bugs:**

- Low battery check on startup [\#13](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/13)

**Closed issues:**

- HW schematic [\#3](https://github.com/AZURTEST-CORP/SSST_blood_sugar/issues/3)

**Merged pull requests:**

- Merge of HW branch for milestone V0 [\#16](https://github.com/AZURTEST-CORP/SSST_blood_sugar/pull/16) ([jernejp21](https://github.com/jernejp21))
- Merge of FW branch for milestone V0 [\#14](https://github.com/AZURTEST-CORP/SSST_blood_sugar/pull/14) ([jernejp21](https://github.com/jernejp21))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


Firmware/src/IO.c
/**
 * SSST blood sugar. Device for contacless measuring of blood sugar.
 * Copyright (C) 2022  Azurtest
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "IO.h"
#include "config.h"
#include <hal/nrf_gpio.h>
#include <hal/nrf_saadc.h>
#include <hal/nrf_spim.h>
#include <hal/nrf_timer.h>

void LED_StatusOn()
{
  nrf_gpio_pin_set(LED_STATUS);
}

void LED_StatusOff()
{
  nrf_gpio_pin_clear(LED_STATUS);
}

void LED_StatusToggle()
{
  nrf_gpio_pin_toggle(LED_STATUS);
}

void LED_BatteryOn()
{
  nrf_gpio_pin_set(LED_BATT);
}

void LED_BatteryOff()
{
  nrf_gpio_pin_clear(LED_BATT);
}

void DCDC_3V3SwitchOff()
{
  nrf_gpio_pin_clear(DCDC_SHTDN_3V3);
}

void DCDC_3V3SwitchOn()
{
  nrf_gpio_pin_set(DCDC_SHTDN_3V3);
}

void DCDC_5VSwitchOff()
{
  nrf_gpio_pin_clear(DCDC_SHTDN_5V);
}

void DCDC_5VSwitchOn()
{
  nrf_gpio_pin_set(DCDC_SHTDN_5V);
}

uint8_t BTN_ShutdnStatus()
{
  uint8_t status = nrf_gpio_pin_read(BTN_SHUTDN);
  return status;
}

uint8_t BTN_AdcChgStatus()
{
  uint8_t status = nrf_gpio_pin_read(BTN_ADC_CHG);
  return status;
}

void SAADC_EnableIntADC()
{
  // Enable SAADC and start it
  nrf_saadc_enable(NRF_SAADC);
  nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_START);
}

void SAADC_DisableIntADC()
{
  // Stop and clear timer
  timerStopSampling();

  // Disable SAADC EasyDMA interrupt
  nrf_saadc_disable(NRF_SAADC);
  nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_STOP);
  nrf_saadc_event_clear(NRF_SAADC, NRF_SAADC_EVENT_STARTED);
  nrf_saadc_event_clear(NRF_SAADC, NRF_SAADC_EVENT_END);
  nrf_saadc_event_clear(NRF_SAADC, NRF_SAADC_EVENT_DONE);
  nrf_saadc_event_clear(NRF_SAADC, NRF_SAADC_EVENT_RESULTDONE);
  nrf_saadc_event_clear(NRF_SAADC, NRF_SAADC_EVENT_STOPPED);
}

void SAADC_EnableExtADC()
{
  // Enable SPI
  nrf_spim_enable(NRF_SPIM3);
}

void SAADC_DisableExtADC()
{
  // Stop and clear timer
  timerStopSampling();

  // Disable SPI EasyDMA interrupt
  nrf_spim_disable(NRF_SPIM3);
  nrf_spim_event_clear(NRF_SPIM3, NRF_SPIM_EVENT_STOPPED);
  nrf_spim_event_clear(NRF_SPIM3, NRF_SPIM_EVENT_ENDRX);
  nrf_spim_event_clear(NRF_SPIM3, NRF_SPIM_EVENT_END);
  nrf_spim_event_clear(NRF_SPIM3, NRF_SPIM_EVENT_ENDTX);
  nrf_spim_event_clear(NRF_SPIM3, NRF_SPIM_EVENT_STARTED);
}

void timerStopSampling()
{
  nrf_timer_task_trigger(NRF_TIMER2, NRF_TIMER_TASK_STOP);
  nrf_timer_task_trigger(NRF_TIMER2, NRF_TIMER_TASK_CLEAR);
  nrf_timer_event_clear(NRF_TIMER2, NRF_TIMER_EVENT_COMPARE0);
  nrf_timer_task_trigger(NRF_TIMER3, NRF_TIMER_TASK_CLEAR);
}

void timerStartSampling()
{
  nrf_timer_task_trigger(NRF_TIMER0, NRF_TIMER_TASK_START);
}


Firmware/src/IO.h
/**
 * SSST blood sugar. Device for contacless measuring of blood sugar.
 * Copyright (C) 2022  Azurtest
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <stdint.h>
#include <stdio.h>

#ifndef __IO_H
#define __IO_H

void LED_StatusOn();
void LED_StatusOff();
void LED_StatusToggle();
void LED_BatteryOn();
void LED_BatteryOff();
void DCDC_3V3SwitchOff();
void DCDC_3V3SwitchOn();
void DCDC_5VSwitchOff();
void DCDC_5VSwitchOn();
uint8_t BTN_ShutdnStatus();
uint8_t BTN_AdcChgStatus();
void SAADC_EnableIntADC();
void SAADC_DisableIntADC();
void SAADC_EnableExtADC();
void SAADC_DisableExtADC();
void timerStopSampling();
void timerStartSampling();

#endif  //__IO_H

winIDEA/OSprofiler.py
import time
import isystem.connect as ic

connectionMgr = ic.ConnectionMgr()
connectionMgr.connectMRU('')

trdFileName = "OS_profiler.trd"

debugConfig = ic.CDebugFacade(connectionMgr)
profileConfig = ic.CProfilerController2(connectionMgr, trdFileName, "w")
ideConfig = ic.CIDEController(connectionMgr)

debugConfig.run()

# Wait that all threads are really started.
time.sleep(3)

eval = debugConfig.evaluate(ic.IConnectDebug.fRealTime, "_kernel.threads")
threadNames = []
threadsAddresses = []

nextThread = "->next_thread"
cnt = 0

while(1):
    eval = debugConfig.evaluate(ic.IConnectDebug.fRealTime, "_kernel.threads" + nextThread * cnt)
    if(eval.getAddress().m_aAddress == 0):
        break
    name = debugConfig.evaluate(ic.IConnectDebug.fRealTime, "_kernel.threads" + nextThread * cnt + "->name")
    nameStr = name.getResult(ic.CValueType.EChar).strip('"')
    threadAddress = eval.getAddress().m_aAddress
    cnt += 1
    threadNames.append(nameStr)
    threadsAddresses.append(threadAddress)


profileConfig.setAnalysisType(0, ic.CProfilerController2.EProfiler)
profileConfig.setProfilingSections(0, isProfileCode=False, isProfileData=True, isProfileOSObjects=False, isProfileAUX=False)

'''
Create header file for thread addresses
'''

with open("thread_names.h", "w") as file:
    for index, name in enumerate(threadNames):
        name = name.upper()
        name = name.replace(" ", "_")
        strToWrite = "#define {} {}\n".format(name.upper(), hex(threadsAddresses[index]))
        file.writelines(strToWrite)

profileConfig.addVariable(0, "_kernel.cpus[0].current", ic.CProfilerController2.EDATState)

description = "/Document/{}/Trigger.Items[0].Processes[0].Profiler.DataArea[0].Description".format(trdFileName)
ideConfig.setOption(description, "CPU_0")

define = "/Document/{}/Trigger.Items[0].Processes[0].Profiler.DataArea[0].Definition".format(trdFileName)
ideConfig.setOption(define, "define")

headerFile = "/Document/{}/Trigger.Items[0].Processes[0].Profiler.DataArea[0].DefinitionLocation".format(trdFileName)
ideConfig.setOption(headerFile, "thread_names.h")

Firmware/src/config.h
/**
 * SSST blood sugar. Device for contacless measuring of blood sugar.
 * Copyright (C) 2022  Azurtest
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef __CONFIG_H
#define __CONFIG_H

/* Defines */
#define RAW_DATA_SIZE 10  // This will change after testing
#define AVG_DATA_SIZE 100  // This will change after testing
#define BTN_SHUTDN_DELAY 2500  // value in ms
#define ADC_RESOLUTION 4096  // 12-bit resolution
#define SAMPLING_RATE 1000  // 1000000/SAMPLING_RATE is samp. freq.
#define PIN_MAP(port, pin) (port * 32 + pin)

/* HW defines. LEDs, Buttons etc. */
#define BTN_SHUTDN PIN_MAP(0, 21)
#define BTN_ADC_CHG PIN_MAP(0, 22)
#define LED_STATUS PIN_MAP(0, 25)
#define LED_BATT PIN_MAP(0, 24)
#define DCDC_SHTDN_3V3 PIN_MAP(0, 13)
#define DCDC_SHTDN_5V PIN_MAP(0, 29)
#define AIN_BATT_SENS PIN_MAP(0, 3)
#define AIN_BLOOD_SENS PIN_MAP(0, 4)
#define SPI_SCK_PIN PIN_MAP(0, 19)
#define SPI_MOSI_PIN PIN_MAP(0, 18)
#define SPI_MISO_PIN PIN_MAP(0, 20)
#define SPI_CS_PIN PIN_MAP(0, 17)

/* HW configuration. LOW, HIGH logic etc. */
//#define BUTTON_POS_LOGIC  // Is button positve logic? Released LOW state, pressed HIGH state
#define SUPPLY_VOLTAGE (4.2)
#define BATT_MIN_VOLT (SUPPLY_VOLTAGE * 12 / 16)  // Not actual voltage, but adjusted where full battery (about 4.2 V) is suppy voltage

#ifdef BUTTON_POS_LOGIC
#define BUTTON_LOGIC 1
#else
#define BUTTON_LOGIC 0
#endif

/* Enums */

enum
{
    ERROR_AMP = 1,
    ERROR_ADC,
    ERROR_OTH,
    ERROR_NONE = 0xFF,
};

#endif  //__CONFIG